src/fsharp/service/SemanticClassification.fs

```diff
@@ -0,0 +1,183 @@
+               (fun () -> 
+                let (|LegitTypeOccurence|_|) = function
+                    | ItemOccurence.UseInType
+                    | ItemOccurence.UseInAttribute
+                    | ItemOccurence.Use _
+                    | ItemOccurence.Binding _
+                    | ItemOccurence.Pattern _ -> Some()
+                    | _ -> None
+
+                let (|OptionalArgumentAttribute|_|) ttype =
+                    match ttype with
+                    | TType.TType_app(tref, _) when tref.Stamp = g.attrib_OptionalArgumentAttribute.TyconRef.Stamp -> Some()
+                    | _ -> None
+
+                let (|KeywordIntrinsicValue|_|) (vref: ValRef) =
+                    if valRefEq g g.raise_vref vref ||
+                       valRefEq g g.reraise_vref vref ||
+                       valRefEq g g.typeof_vref vref ||
+                       valRefEq g g.typedefof_vref vref ||
+                       valRefEq g g.sizeof_vref vref ||
+                       valRefEq g g.nameof_vref vref
+                    then Some()
+                    else None
+            
+                let (|EnumCaseFieldInfo|_|) (rfinfo : RecdFieldInfo) =
+                    match rfinfo.TyconRef.TypeReprInfo with
+                    | TFSharpObjectRepr x ->
+                        match x.fsobjmodel_kind with
+                        | TTyconEnum -> Some ()
+                        | _ -> None
+                    | _ -> None
+
+                let resolutions =
+                    match range with
+                    | Some range ->
+                        sResolutions.CapturedNameResolutions
+                        |> Seq.filter (fun cnr -> rangeContainsPos range cnr.Range.Start || rangeContainsPos range cnr.Range.End)
+                    | None -> 
+                        sResolutions.CapturedNameResolutions :> seq<_>
+
+                let isDisposableTy (ty: TType) =
+                    protectAssemblyExplorationNoReraise false false (fun () -> Infos.ExistsHeadTypeInEntireHierarchy g amap range0 ty g.tcref_System_IDisposable)
+
+                let isStructTyconRef (tyconRef: TyconRef) = 
+                    let ty = generalizedTyconRef tyconRef
+                    let underlyingTy = stripTyEqnsAndMeasureEqns g ty
+                    isStructTy g underlyingTy
+
+                let isValRefMutable (vref: ValRef) =
+                    // Mutable values, ref cells, and non-inref byrefs are mutable.
+                    vref.IsMutable
+                    || Tastops.isRefCellTy g vref.Type
+                    || (Tastops.isByrefTy g vref.Type && not (Tastops.isInByrefTy g vref.Type))
+
+                let isRecdFieldMutable (rfinfo: RecdFieldInfo) =
+                    (rfinfo.RecdField.IsMutable && rfinfo.LiteralValue.IsNone)
+                    || Tastops.isRefCellTy g rfinfo.RecdField.FormalType
+
+                let duplicates = HashSet<range>({ new IEqualityComparer<range> with 
```

Comment:
I think you could add a `Range.comparer : IEqualityComparer` to the range module.

Reply:
Makes sense.